### ISSUE TITLE
Vehicle characteristics scraper

### ISSUE THREAD

**Issue by tprjd** [X-rays5/wt-api#1]
Hey!

I had an idea for a while to make a tool where you can filter vehicles by br and every other metric. For example you select a tank and you can view what tanks can it face etc.
Your work is excellent but I would need a scraper for vehicle characteristics as well. I would love to contribute but I don't know Rust.
What do you think how much work would it be to do the vehicle scraper?

**Comment by X-rays5** [comment]
Hi,

I originally planned on scraping the vehicle pages but never did since i didn't end up needing it. The only problem with scraping the vehicle pages is. That it can't do them all at once since the execution time limit on cloudflare workers is too short. But that is easily solved by running that part on a different server. But that may cause a rate limit since there are a lot of vehicles. Although i have never tested this.

Other then that if i remember correctly the vehicles pages html is quite complex. So it might be a bit harder to scrape all the data from it. To cache all of that correctly. The current way of how data is cached in the DB might need to change a bit. But that's fine since that's a mess anyways. I got really confused how i made that when I was trying to fix something.

TL;DR
Scraping the vehicle pages should be doable. If a request is up to 5 vehicles at a time or something around that. Would need to test what a good limit would be. Might also need to rethink how to store the cache in the DB.

**Comment by tprjd** [comment]
It should be enough to gather the data once every patch or something. 
Now I checked out a few vehicle pages and yes it really is complex. :( 
Also would the free tier be enough on heroku to store everything?

**Comment by X-rays5** [comment]
There are currently about 2400 vehicles in game. So assuming that each vehicle has 64kb of data (which is probably way too much). All the data will only take 154MB. But first thing to do is to check what the wiki rate limit is. 

It also should be possible to use the steam api to check for a new patch.

**Comment by tprjd** [comment]
Hey! It's been a while. :)
So I did some scraping on my own using you apis vehicle list as a base. 
There are some vehicles that has wrong names and wiki link. The `_`s are missing. eg: 
```
{
     "name": "GladiatorMkII",
     "thumbnail_img_url": "https://encyclopedia.warthunder.com/slots/gladiator_mk2.png",
     "wiki_page": "https://wiki.warthunder.com/GladiatorMkII"
}
```
it should be `Gladiator_Mk_II`

This is where I'm at currently if interested:
https://wt-portal.vercel.app/vehicles/filter

**Comment by X-rays5** [comment]
Hey,
I took a look for you. It seems that some vehicles like the Gladiator use nbsp instead of a normal space. This breaks the way that i get the wiki url. But the way I'm getting that url is kinda stupid anyways. I'll quickly look into it and try to fix it.

**Comment by X-rays5** [comment]
Hi,
36a914d Should fix your issue. I tested some urls for random vehicles and they all seem to still work. Let me know if you have any further issues. This is the new output:
```json
{
    "name": "Gladiator Mk II",
    "thumbnail_img_url": "https://encyclopedia.warthunder.com/slots/gladiator_mk2.png",
    "wiki_page": "https://wiki.warthunder.com/Gladiator_Mk_II"
 }
```

Also site looks cool. Interested in seeing how it will be when fully functional.

**Comment by tprjd** [comment]
Hey

Sorry to inform you but I'm not using your api anyone nor do I scrape anything. I found out that the [datamine repo](https://github.com/gszabi99/War-Thunder-Datamine) contains everything as a json. It's still hard to parse but at least somewhat consistent.
And the current progress is: [Techtree](https://wt-portal.vercel.app/techtree) :)

**Comment by X-rays5** [comment]
Hey

Looks pretty cool. Especially the raw json button is quite handy. Don't need to search trough the files to get that anymore. Only thing I'm wondering is where that raw json comes from. Since it doesn't seem to be the contents of the blkx file.

And no worries about not using the api anymore. The only reason it even exists is because i wanted to try web scraping with rust. Getting the data from the datamine repo is a much better idea honestly.
